allennlp_models/pair_classification/dataset_readers/snli.py

```diff
@@ -60,20 +62,29 @@ def _read(self, file_path: str):
         # if `file_path` is a URL, redirect to the cache
         file_path = cached_path(file_path)
 
-        with open(file_path, "r") as snli_file:
+        import torch.distributed as dist
+        from allennlp.common.util import is_distributed
+
+        if is_distributed():
+            start_index = dist.get_rank()
+            step_size = dist.get_world_size()
+            logger.info(
+                "Reading SNLI instances %% %d from jsonl dataset at: %s", step_size, file_path
+            )
+        else:
+            start_index = 0
+            step_size = 1
             logger.info("Reading SNLI instances from jsonl dataset at: %s", file_path)
-            for line in snli_file:
-                example = json.loads(line)
 
+        with open(file_path, "r") as snli_file:
+            example_iter = (json.loads(line) for line in snli_file)
+            filtered_example_iter = (
+                example for example in example_iter if example["gold_label"] != "-"
+            )
+            for example in itertools.islice(filtered_example_iter, start_index, None, step_size):
```

Comment:
You could do the `islice` before the `json.loads`, so that it doesn't have to parse the JSON for lines that get discarded anyways. The problem is that then the filtering has to happen afterwards, which would mean different workers have different number of instances. This is not ideal, but not the end of the world.

I'm just leaving this here as a suggestion.

Reply:
I had it that way initially but it didn't make a significant impact on speed.